File: app/src/main/java/com/example/federico/mlibrefedericopuy/ui/ProductDetailActivity.java
package com.example.federico.mlibrefedericopuy.ui;

import android.content.Intent;
import android.os.Bundle;
import android.support.v7.widget.Toolbar;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.app.ActionBar;
import android.view.MenuItem;

import com.example.federico.mlibrefedericopuy.R;
import com.example.federico.mlibrefedericopuy.utils.Constants;

/**
 * Actividad usada unicamente para celulares o dispositivos con pantallas chicas
 */
public class ProductDetailActivity extends AppCompatActivity {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_product_detail);
        Toolbar toolbar = (Toolbar) findViewById(R.id.detail_toolbar);
        setSupportActionBar(toolbar);

        ActionBar actionBar = getSupportActionBar();
        if (actionBar != null) {
            actionBar.setDisplayHomeAsUpEnabled(true);
        }

        if (savedInstanceState == null) {

            String stepJson = getIntent().getStringExtra(Constants.PRODUCT_INTENT_JSON);
            Bundle arguments = new Bundle();
            arguments.putString(Constants.PRODUCT_FRAGMENT_JSON, stepJson);
            ProductDetailFragment fragment = new ProductDetailFragment();
            fragment.setArguments(arguments);
            getSupportFragmentManager().beginTransaction()
                    .add(R.id.item_detail_container, fragment)
                    .commit();
        }
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int id = item.getItemId();
        if (id == android.R.id.home) {
            navigateUpTo(new Intent(this, ProductListActivity.class));
            return true;
        }
        return super.onOptionsItemSelected(item);
    }
}

File: app/src/main/java/com/example/federico/mlibrefedericopuy/detail/ProductDetailActivity.kt
package com.example.federico.mlibrefedericopuy.detail


import android.content.Intent
import android.os.Bundle
import android.view.MenuItem
import android.view.View
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.Observer
import com.example.federico.mlibrefedericopuy.R
import com.example.federico.mlibrefedericopuy.core.getViewModel
import com.example.federico.mlibrefedericopuy.list.ProductListActivity
import com.example.federico.mlibrefedericopuy.model.Description
import com.example.federico.mlibrefedericopuy.model.ProductInfo
import com.example.federico.mlibrefedericopuy.utils.Constants
import kotlinx.android.synthetic.main.activity_product_detail.*
import java.util.*

class ProductDetailActivity : AppCompatActivity() {

    lateinit var viewModel: ProductDetailViewModel

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_product_detail)
        setSupportActionBar(detail_toolbar)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)

        val product = intent.getParcelableExtra<ProductInfo>(Constants.PRODUCT_INTENT_JSON)
        renderProductInfo(product)

        viewModel = getViewModel { ProductDetailViewModel(ProductDetailRepository(), product.id) }

        viewModel.productDescription.observe(this, Observer(this::renderProductDescription))
        viewModel.productImages.observe(this, Observer(this::loadProductImages))
        viewModel.loadingStatus.observe(this, Observer(this::showLoading))

    }

    private fun renderProductInfo(product: ProductInfo) {
        tvTitle.text = product.title
        tvPrice.text = product.price
        tvAvailableQuantity.text = product.availableQuantity
        tvAmountSold.text = product.soldQuantity
        tvCondition.text = product.condition
        if (product.acceptsMp) {
            logoMpago.visibility = View.VISIBLE
        }
    }

    private fun renderProductDescription(description: Description?) {
        if (!description?.plainText.isNullOrBlank()) {
            descriptionTitle.visibility = View.VISIBLE
            tvDescription.visibility = View.VISIBLE
            tvDescription.text = description?.plainText
        }
    }

    private fun loadProductImages(pictures: List<String>?) {
        val productDetailImagesPagerAdapter = ProductDetailImagesPagerAdapter(this, pictures as ArrayList<String>)
        imageViewPager.adapter = productDetailImagesPagerAdapter
    }

    private fun showLoading(isLoading: Boolean?) {
        isLoading?.let {
            if (it) progressBar.visibility = View.VISIBLE else progressBar.visibility = View.INVISIBLE
        }
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        val id = item.itemId
        if (id == android.R.id.home) {
            navigateUpTo(Intent(this, ProductListActivity::class.java))
            return true
        }
        return super.onOptionsItemSelected(item)
    }


}
